sol/main.c

```diff
@@ -0,0 +1,478 @@
+	 * does not currently exist, and there is at least one
+	 * request of a higher priority already in the queue.
+	 */
+	RTE_VERIFY(priority != req_queue->highest_priority);
+	for (next = priority + 1; next <= req_queue->highest_priority; next++) {
+		if (req_queue->priorities[next] != NULL) {
+			list_add(&pr->list, &req_queue->priorities[next]->list);
+			return;
+		}
+	}
+
+	rte_panic("sol: %s failed to insert a request of a new priority\n",
+		__func__);
+}
+
+static void
+drop_lowest_priority_pkt(struct sol_config *sol_conf,
+	struct req_queue *req_queue)
+{
+	struct priority_req *lowest_pr = list_last_entry(&req_queue->head,
+		struct priority_req, list);
+	struct priority_req *next_lowest_pr;
+
+	RTE_VERIFY(req_queue->len > 0);
+
+	if (unlikely(req_queue->len == 1)) {
+		req_queue->priorities[lowest_pr->priority] = NULL;
+		req_queue->highest_priority = 0;
+		req_queue->lowest_priority = GK_NUM_REQ_PRIORITIES;
+		goto drop;
+	}
+
+	next_lowest_pr = list_prev_entry(lowest_pr, list);
+
+	/* The lowest priority packet was the only one of that priority. */
+	if (lowest_pr->priority != next_lowest_pr->priority) {
+		req_queue->priorities[lowest_pr->priority] = NULL;
+		req_queue->lowest_priority = next_lowest_pr->priority;
+		goto drop;
+	}
+
+	req_queue->priorities[lowest_pr->priority] = next_lowest_pr;
+
+drop:
+	list_del(&lowest_pr->list);
+	rte_pktmbuf_free(lowest_pr->pkt);
+	mb_free_entry(&sol_conf->mb, lowest_pr);
+	req_queue->len--;
+}
+
+static void
+enqueue_req(struct sol_config *sol_conf, struct priority_req *req)
+{
+	struct req_queue *req_queue = &sol_conf->req_queue;
+	uint8_t priority = req->priority;
+
+	if (req_queue->len == sol_conf->pri_req_max_len) {
+		/* New packet is lowest priority, so drop it. */
+		if (req_queue->lowest_priority >= priority) {
+			rte_pktmbuf_free(req->pkt);
+			mb_free_entry(&sol_conf->mb, req);
+			return;
+		}
+		drop_lowest_priority_pkt(sol_conf, req_queue);
+	}
+
+	if (req_queue->priorities[priority] == NULL) {
+		/* Insert request of a priority we don't yet have. */
+		insert_new_priority_req(req_queue, req, priority);
+	} else {
+		/* Append request to end of the appropriate priority. */
+		list_add(&req->list, &req_queue->priorities[priority]->list);
+		req_queue->priorities[priority] = req;
+	}
+
+	req_queue->len++;
+}
+
+static void
+enqueue_reqs(struct sol_config *sol_conf)
+{
+	struct priority_req *req_nodes[sol_conf->enq_burst_size];
+	int num_reqs = mb_dequeue_burst(&sol_conf->mb,
+		(void **)req_nodes, sol_conf->enq_burst_size);
+	int i;
+	for (i = 0; i < num_reqs; i++)
+		enqueue_req(sol_conf, req_nodes[i]);
+}
+
+static inline void
+credits_update(struct req_queue *req_queue)
+{
+	uint64_t curr_cycles = rte_get_tsc_cycles();
+	uint64_t avail_cycles = req_queue->spare_cycles +
+		(curr_cycles - req_queue->time_cpu_cycles);
+	uint64_t avail_bytes;
+
+	/* Not enough cycles have passed to update the number of credits. */
+	/* XXX We should run tests to see if this is (un)likely. */
```

Comment:
We can drop this `XXX` comment because not only does the probability depend on our code, but also on how fast the processor and NICs are.